packages/vite/src/node/plugins/dynamicImportVars.ts

```diff
@@ -0,0 +1,242 @@
+import { parseRequest } from '../utils'
+import { parse as parseJS } from 'acorn'
+import { createFilter } from '@rollup/pluginutils'
+import { dynamicImportToGlob } from '@rollup/plugin-dynamic-import-vars'
+import type { TransformGlobImportResult } from './importMetaGlob'
+import { transformGlobImport } from './importMetaGlob'
+import type { ViteDevServer } from '../server'
+
+interface DynamicImportRequest {
+  as?: 'raw'
+}
+
+interface DynamicImportPattern {
+  globParams: DynamicImportRequest | null
+  userPattern: string
+  rawPattern: string
+}
+
+const dynamicImportHelper = (glob: Record<string, any>, path: string) => {
+  const v = glob[path]
+  if (v) {
+    return typeof v === 'function' ? v() : Promise.resolve(v)
+  }
+  return new Promise((_, reject) => {
+    ;(typeof queueMicrotask === 'function' ? queueMicrotask : setTimeout)(
+      reject.bind(null, new Error('Unknown variable dynamic import: ' + path))
+    )
+  })
+}
+export const dynamicImportHelperId = '/@vite/dynamic-import-helper'
+
+export function parseDynamicImportPattern(
+  strings: string
+): DynamicImportPattern | null {
+  const filename = strings.slice(1, -1)
+  const rawQuery = parseRequest(filename)
+  let globParams: DynamicImportRequest | null = null
+  const ast = (
+    parseJS(strings, {
+      ecmaVersion: 'latest',
+      sourceType: 'module'
+    }) as any
+  ).body[0].expression
+
+  const userPatternQuery = dynamicImportToGlob(ast, filename)
+  if (!userPatternQuery) {
+    return null
+  }
+
+  const [userPattern] = userPatternQuery.split('?', 2)
+  const [rawPattern] = filename.split('?', 2)
+
+  if (rawQuery?.raw !== undefined) {
+    globParams = { as: 'raw' }
+  }
+
+  return {
+    globParams,
+    userPattern,
+    rawPattern
+  }
+}
+
+export async function transformDynamicImportGlob(
+  source: string,
+  root: string,
+  importer: string,
+  start: number,
+  end: number,
+  resolve: (url: string, importer?: string) => Promise<string | undefined>
+): Promise<{
+  glob: TransformGlobImportResult
+  pattern: string
+  rawPattern: string
+} | null> {
+  let fileName = source.slice(start, end)
+
+  if (fileName[1] !== '.' && fileName[1] !== '/' && resolve) {
+    const resolvedFileName = await resolve(fileName.slice(1, -1), importer)
+    if (!resolvedFileName) {
+      return null
+    }
+    const relativeFileName = path.posix.relative(
+      path.dirname(importer),
+      resolvedFileName
+    )
+    fileName = `\`${relativeFileName}\``
+  }
+
+  const dynamicImportPattern = parseDynamicImportPattern(fileName)
+  if (!dynamicImportPattern) {
+    return null
+  }
+  const { globParams, rawPattern, userPattern } = dynamicImportPattern
+  const params = globParams ? `, ${JSON.stringify(globParams)}` : ''
+  const exp = `import.meta.glob(${JSON.stringify(userPattern)}${params})`
+  const glob = await transformGlobImport(
+    exp,
+    importer,
+    root,
+    resolve,
+    false,
+    false
+  )
```

Comment:
I am not sure if this needs to be expanded right away. Can't we just return `exp` and let the glob plugin handle this?

Reply:
`importGlobPlugin` is earlier than `importAnalysisPlugin`, `importAnalysisPlugin` should need to handle `import('@/${a}.js')`

Reply:
Should we always do this rewriting even in dev for consistency so we don't have this problem?

Reply:
`importGlobPlugin` is earlier than `importAnalysisPlugin`.
so dev will got this unhandled expression
```js
__variableDynamicImportRuntimeHelper(import.meta.glob('./*.js'))
```